gprofiler/metadata/py_package_version.py

```diff
@@ -0,0 +1,258 @@
+       from ``info``; if ``info`` is empty, start appending ``..`` instead.
+    2. Join the two directly.
+    """
+    while entry and entry[0] == "..":
+        if not info or info[-1] == "..":
+            info += ("..",)
+        else:
+            info = info[:-1]
+        entry = entry[1:]
+    return str(pathlib.Path(*info, *entry))
+
+
+def _files_from_record(dist: pkg_resources.Distribution) -> Optional[Iterator[str]]:
+    """Based on _files_from_record in pip._internal.commands.show.search_packages_info"""
+    try:
+        text = dist.get_metadata("RECORD")
+    except FileNotFoundError:
+        return None
+    # This extra Path-str cast normalizes entries.
+    return (str(pathlib.Path(row[0])) for row in csv.reader(text.splitlines()))
+
+
+def _files_from_legacy(dist: pkg_resources.Distribution) -> Optional[Iterator[str]]:
+    """Based on _files_from_legacy in pip._internal.commands.show.search_packages_info"""
+    try:
+        text = dist.get_metadata("installed-files.txt")
+    except FileNotFoundError:
+        return None
+    paths = (p for p in text.splitlines(keepends=False) if p)
+    root = dist.location
+    info = dist.egg_info
+    if root is None or info is None:
+        return paths
+    try:
+        info_rel = pathlib.Path(info).relative_to(root)
+    except ValueError:  # info is not relative to root.
+        return paths
+    if not info_rel.parts:  # info *is* root.
+        return paths
+    return (_convert_legacy_entry(pathlib.Path(p).parts, info_rel.parts) for p in paths)
+
+
+def _get_package_name(dist: pkg_resources.Distribution) -> Optional[str]:
+    """Based on pip._internal.metadata.base.BaseDistribution.raw_name"""
+    # TODO: Test
+    metadata = _get_metadata(dist)
+    if metadata:
+        # The metadata should NEVER be missing the Name: key, but if it somehow
+        # does, fall back to the known canonical name.
+        return metadata.get("Name", dist.project_name)
+    return None
+
+
+def _get_libpython_path(pid: int) -> Optional[str]:
+    libpython_maps_pattern = re.compile(r"(?<=\s)/\S*libpython\S*\.so(\.\S+)?\Z")
+    with open(f"/proc/{pid}/maps") as f:
+        for line in f.readlines():
+            match = libpython_maps_pattern.search(line.strip())
+            if match is not None:
+                return match.group()
+    return None
+
+
+@functools.lru_cache(maxsize=128)
+def _get_python_full_version(pid: int, short_version: str) -> Optional[str]:
+    assert re.match(r"[23]\.\d\d?", short_version)
+
+    bin_file = _get_libpython_path(pid) or f"/proc/{pid}/exe"
+    full_version_string_pattern = re.compile(rb"(?<=\D)" + short_version.encode() + rb"\.\d\d?(?=\x00)")
+
+    # Try to extract the version string from the binary
+    with open(convert_to_proc_root_path(bin_file, pid), "rb") as f:
+        for line in f.readlines():
+            match = full_version_string_pattern.search(line)
+            if match is not None:
+                return match.group().decode()
```

Comment:
Why splitting lines? Can just match against the whole bytes buffer

Reply:
I had in mind that I don't want to load the entire file at once because I was concerned about memory usage. Now that I think about it I'm not sure that `file.read()` actually loads the entire file like that.

Reply:
I'm aware that it's not a textual file. `readlines` is used here as an arbitrary way to split the file in a way that ensures that we won't cut `PY_VERSION`.
I'll change it though.